### PR TITLE
feat!: add Metadata management

### PR DESCRIPTION
**Pull request by Slesarew**
### Major features

- [x] Metadata is stored dynamically in secure storage database
- [x] Arbitrary transactions could be parsed and previewed based on stored metadata
- [x] Used metadata version is shown on network settings screen
- [x] Metadata could be changed in network settings
- [x] New metadata could be uploaded through animated series of fountain QR codes
- [x] Built-in metadata list is updated dynamically through automation
- [x] Metadata could be permanently deleted from the device (except built-in versions)
- [ ] Add tests for new features
- [x] Rust code was moved back into this repo

### Optional features:

- [ ] Rust transaction parser
- [ ] Pretty transaction payload rendering
- [ ] Fix weird RN error messages on hook mixup
- [x] Add "wait for loading" notifications
- [x] Add notification upon successful metadata addition that asks to verify hash and confirms download
- [ ] Fix CI
- [ ] Integrate both scanners into one screen

## This PR would probably address the following issues:

#407
#318 (patrially until https://github.com/paritytech/substrate/pull/8615)
#408
#758
#457 (partially)
#713
#724
#736
#739
#767

Metadata version and name are parsed from metadata, thus networks that do not store it in system constants are not fully supported here - they are identified only by their hash and should be used with great care.

### Metadata hashes

The metadata hashes are generated as 32 bytes blake2b with no password. They are used for 3 purposes:

1. Metadata validation (through manual hash comparison now, some automatic tools and visualization should be developed, I'll open an issue soon)
2. Identification of versionless metadata for storage
3. Internal hashtables

### CODE REVIEW

File: rust/signer/src/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+pub fn get_meta_const (meta: &str) -> Result<Vec<u8>, &str> {
+
+    if !meta.starts_with("0x6d657461") {
+        return Err("No 'meta' starting sequence in metadata");
+    }
+    
+    let part1 = &meta[10..12];
+    let part1_vec = hex::decode(part1).unwrap();
+    let part1_decoded = u8::decode(&mut &part1_vec[..]).unwrap();
+    if part1_decoded != 12 {
```
Comment (emostov) [inline rust/signer/src/metadata.rs:34]: Does this mean when v13 comes out we no longer will be able to update the metadata?

File: rust/signer/src/metadata.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
+use hex;
```
Comment (emostov) [inline rust/signer/src/metadata.rs:1]: Whats the advantage of rolling up one-off metadata processing vs using @polkadot/metadata?

File: src/modules/main/components/NetworkSelector.tsx [hunk 3]
```diff
@@ -60,6 +60,7 @@ function NetworkSelector({
 
 	const { setAlert } = useContext(AlertStateContext);
 	// catch android back button and prevent exiting the app
+	// TODO: this just doesn't work and nobody noticed, let's fix later
```
Comment (emostov) [inline src/modules/main/components/NetworkSelector.tsx:63]: GitHub issue?

File: src/modules/sign/components/PayloadDetailsCard.tsx [hunk 4]
```diff
@@ -58,72 +78,70 @@ const ExtrinsicPart = withRegistriesStore<ExtrinsicPartProps>(
...
-						if (!meta.args.length) {
-							const sectionMethod = `${call.method}.${call.section}`;
-							callMethodArgs[sectionMethod] = null;
-							return;
-						}
+				return argument;
+			});
+		}
+
+		function parseGenericCall(genericCall: GenericCall): SanitizedCall {
```
Comment (emostov) [inline src/modules/sign/components/PayloadDetailsCard.tsx:101]: Should we have some recursion depth limiter?

File: rust/signer/src/metadata.rs (hunk 2 above)
Comment (Slesarew) [reply rust/signer/src/metadata.rs:1]: noticeably faster

File: src/stores/NetworkContext.ts [hunk 5]
```diff
@@ -70,14 +146,92 @@ export function useNetworksContext(): NetworksContextState {
...
+					console.log(networkParams.pathId);
+					const metadataHandle = networkParams.metadata;
+					const networkMetadataRaw = await getMetadata(metadataHandle);
+					const newRegistry = new TypeRegistry();
+					const overrideTypes = getOverrideTypes(
+						newRegistry,
+						networkParams.pathId
+					);
+					//const overrideTypes = getSpecTypes(newRegistry, networkParams.pathId, metadataHandle.specName, Number.MAX_SAFE_INTEGER);
+					newRegistry.register(overrideTypes);
```
[the remark this answers is not in the document]
Comment (Slesarew) [reply src/stores/NetworkContext.ts:172]: Oh my, this is one of the things I was afraid of  - really difficult to catch before QA team. Thanks, I need to figure this out (actually maybe start rolling out my rust parser code).

File: rust/signer/src/metadata.rs (hunk 1 above)
Comment (Slesarew) [reply rust/signer/src/metadata.rs:34]: > Does this mean when v13 comes out we no longer will be able to update the metadata?

Actually, as far as I understand, losing compatibility on major API change is safety feature here, but confirming this statement wait until our secure interface audit next month.

File: ios/NativeSigner/SubstrateSign-Bridging-Header.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2015-2020 Parity Technologies (UK) Ltd.
```
Comment (goldsteinsveta) [inline ios/NativeSigner/SubstrateSign-Bridging-Header.h:1]: license is outdated

**Comment by emostov** [comment]
> Metadata version and name are parsed from metadata, thus networks that do not store it in system constants are not fully supported here - they are identified only by their hash and should be used with great care.

By hash you mean genesis hash?

**Comment by Slesarew** [comment]
> > Metadata version and name are parsed from metadata, thus networks that do not store it in system constants are not fully supported here - they are identified only by their hash and should be used with great care.
> 
> By hash you mean genesis hash?

No, it is not genesis hash, one metadata could be re-used by different networks. It is metadata hex string's hash - this is subject to discussion now, especially the algorithm used and length. I probably need some more elaborate explanations here, will add them to PR comment so it's simple to find.

**Review by edaj965 — APPROVED** [review]

**Comment by Slesarew** [comment]
Redundant with #791
Forked into named legacy branch legacy-metadataRN

**Pull request closed without merging**